x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -0,0 +1,142 @@
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.xpack.core.XPackSettings;
+import org.elasticsearch.xpack.core.security.action.CreateApiKeyRequest;
+import org.elasticsearch.xpack.core.security.action.CreateApiKeyResponse;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authc.support.Hasher;
+import org.elasticsearch.xpack.security.support.SecurityIndexManager;
+
+import java.io.IOException;
+import java.time.Clock;
+import java.time.Instant;
+import java.util.Arrays;
+
+import static org.elasticsearch.xpack.core.ClientHelper.SECURITY_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+public class ApiKeyService {
+
+    private static final Logger LOGGER = LogManager.getLogger(ApiKeyService.class);
+    private static final String TYPE = "doc";
+
+    private final Clock clock;
+    private final Client client;
+    private final SecurityIndexManager securityIndex;
+    private final ClusterService clusterService;
+    private final boolean enabled;
+
+    public ApiKeyService(Settings settings, Clock clock, Client client,
+                         SecurityIndexManager securityIndex, ClusterService clusterService) {
+        this.clock = clock;
+        this.client = client;
+        this.securityIndex = securityIndex;
+        this.clusterService = clusterService;
+        this.enabled = XPackSettings.API_KEY_SERVICE_ENABLED_SETTING.get(settings);
+    }
+
+    /**
+     * Asynchronously creates a new API key based off of the request and authentication
+     * @param authentication the authentication that this api key should be based off of
+     * @param request the request to create the api key included any permission restrictions
+     * @param listener the listener that will be used to notify of completion
+     */
+    public void createApiKey(Authentication authentication, CreateApiKeyRequest request, ActionListener<CreateApiKeyResponse> listener) {
+        ensureEnabled();
+        if (authentication == null) {
+            listener.onFailure(new IllegalArgumentException("authentication must be provided"));
+        } else {
+            final Instant created = clock.instant();
+            final Instant expiration = getApiKeyExpiration(created, request);
+            final SecureString apiKey = UUIDs.randomBase64UUIDSecureString();
+            final Version version = clusterService.state().nodes().getMinNodeVersion();
+            if (version.before(Version.V_7_0_0_alpha1)) { // TODO(jaymode) change to V6_5_0 on backport!
+                LOGGER.warn("nodes prior to the minimum supported version for api keys {} exist in the cluster; these nodes will not be " +
+                    "able to use api keys", Version.V_7_0_0_alpha1);
+            }
+
+            final char[] keyHash = Hasher.PBKDF2.hash(apiKey);
```

Comment:
Having the cost factor as a configuration parameter would allow users to adjust without us modifying the code. ( This is what I was going for with the initial "for the sake of configuration only future-proofing against brute force attacks" comment).